torch/_inductor/compile_fx.py

```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+            for out in outputs:
+                if out is not None and out() is not None:
+                    return False
+
+        live_outputs = []
+        final_recorded_liveness = self.recorded_liveness_after_graph[-1]
+        for i in range(len(final_recorded_liveness)):
+            for j in range(len(final_recorded_liveness[i])):
+                if final_recorded_liveness[i][j]:
+                    live_outputs.append((i, j))
+
+        self.outputs_live_at_last_tape = live_outputs
+        return True
+
+    def is_cuda_graph_recorded_tensor(self, tensor):
+        cuda_graph_managed_tensors = []
+        for i in range(len(self.recorded_outputs_weakrefs)):
+            for j in range(len(self.recorded_outputs_weakrefs[i])):
+                output = self.recorded_outputs_weakrefs[i][j]
+                if output is not None and output() is tensor:
+                    return True
+
+        return False
+
+
+class TapeManagerContainer(object):
+    """
+    Manages the lifetime of the tape manager. Like `PrivatePool` in cuda caching allocator,
+    the tape and its corresponding memory pool should be kept alive as long as any outstanding
+    graph or tensor which is an output of a graph remains alive.
+    """
+
+    def __init__(self):
+        self.tape_manager = None
+        self.alive_cudagraphify_objs = 0
+
+        # only set in the case where the CudaGraphify objects die while their
+        # tensor outputs remain live
+        self.live_tensors_count = 0
+        # Used to keep alive the memory pool if the CudaGraph object dies
+        self.graph = None
+        self.lock = threading.Lock()
+
+    def finalize_tensor(self):
+        with self.lock:
+            self.live_tensors_count -= 1
+            if self.live_tensors_count == 0:
+                self.tape_manager = None
+                self.graph = None
+
+    def finalize_cuda_graphify(self, graph: torch.cuda.CUDAGraph, tape_manager):
+        with self.lock:
+            self.alive_cudagraphify_objs -= 1
+            if self.alive_cudagraphify_objs != 0:
+                return
+            live_tensors = tape_manager.live_cuda_graph_managed_tensors()
+            if not live_tensors:
+                self.tape_manager = None
+                return
+
+            self.graph = graph
+            for t in live_tensors:
+                self.live_tensors_count += 1
+                weakref.finalize(t, self.finalize_tensor)
+
+    def get_tape_manager(self, obj: Optional["CudaGraphify"] = None):
+        # if the live tensors count is set, then the memory pool is still
+        # being used even if the tape manager and CudaGraphTape objects are dead
+        if self.live_tensors_count:
+            return None
+
+        if self.tape_manager is None or self.tape_manager() is None:
+            tape_manager = CUDAGraphTapeManger()
+            self.tape_manager = weakref.ref(tape_manager)
+        else:
+            tape_manager = self.tape_manager()
```

Comment:
This is a race, the tape manager could die in between when you test it and when you use it.